src/tools/rbd_mirror/ImageReplayer.cc

```diff
@@ -898,13 +898,14 @@ void ImageReplayer<I>::shut_down(int r) {
   // chain the shut down sequence (reverse order)
   Context *ctx = new LambdaContext(
     [this, r](int _r) {
-      update_mirror_image_status(true, STATE_STOPPED);
       handle_shut_down(r);
     });
 
   // destruct the state builder
   if (m_state_builder != nullptr) {
     ctx = new LambdaContext([this, ctx](int r) {
+      set_state_description(0, "local image is primary");
```

Comment:
I agree on your point.

Actually:
```
void ImageReplayer<I>::handle_replayer_notification() {
[...]
  if (!m_replayer->is_replaying()) {
  ┆ auto error_code = m_replayer->get_error_code();
  ┆ auto error_description = m_replayer->get_error_description();
  ┆ dout(10) << "replay interrupted: "
  ┆ ┆ ┆ ┆ ┆ ┆<< "r=" << error_code << ", "
  ┆ ┆ ┆ ┆ ┆ ┆<< "error=" << error_description << dendl;
    on_stop_journal_replay(error_code, error_description);
  ┆ return;
  }
[...]
}

void ImageReplayer<I>::on_stop_journal_replay(int r, const std::string &desc) {
{
  dout(10) << dendl;

  {
  ┆ std::lock_guard locker{m_lock};
  ┆ if (m_state != STATE_REPLAYING) {
  ┆ ┆ // might be invoked multiple times while stopping
  ┆ ┆ return;
  ┆ }

  ┆ m_stop_requested = true;
  ┆ m_state = STATE_STOPPING;
  }

  cancel_update_mirror_image_replay_status();
  set_state_description(r, desc);          ---> Set the state description
  update_mirror_image_status(true, boost::none);
  shut_down(0);                            ---> Invoke shut_down stuff
}
```

I didn't find a better place to `set_state_description`.

Reply:
Well there is an another BZ: https://bugzilla.redhat.com/show_bug.cgi?id=2188205 which says in this case the description should be set to "description: force promoted".
May be we should set a global flag in case of force promote only and based on it set the `set_state_description(0, "force promoted");` in the shut_own() conditionally ?

Reply:
There already is a place in `Replayer::scan_local_mirror_snapshots()` where forced promotion is detected and error description is set to `force promoted`.  If it's not working, let's figure out why.

Reply:
Yeah, I'm actually on it :-)